x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,244 @@
+        final String apiKey = getApiKey(user, password);
+        final Tuple<List<String>, String> httpInfo = getNodeInfo(user, password);
+
+        try {
+            final XContentBuilder builder = JsonXContent.contentBuilder();
+            builder.startObject();
+            builder.field("ver", httpInfo.v2());
+            builder.startArray("adr");
+            for (String bound_address : httpInfo.v1()) {
+                builder.value(bound_address);
+            }
+            builder.endArray();
+            builder.field("fgr", fingerprint);
+            builder.field("key", apiKey);
+            builder.endObject();
+            final String jsonString = Strings.toString(builder);
+            return Base64.getUrlEncoder().encodeToString(jsonString.getBytes(StandardCharsets.UTF_8));
+        } catch (Exception e) {
+            logger.error(("Error generating enrollment token"), e);
+            throw new IllegalStateException("Error generating enrollment token: " + e.getMessage());
+        }
+    }
+
+    private HttpResponse.HttpResponseBuilder responseBuilder(InputStream is) throws IOException {
+        final HttpResponse.HttpResponseBuilder httpResponseBuilder = new HttpResponse.HttpResponseBuilder();
+        if (is != null) {
+            byte[] bytes = toByteArray(is);
+            String responseBody = new String(bytes, StandardCharsets.UTF_8);
+            logger.debug(responseBody);
+            httpResponseBuilder.withResponseBody(responseBody);
+        } else {
+            logger.debug("<Empty response>");
+        }
+        return httpResponseBuilder;
+    }
+
+    private byte[] toByteArray(InputStream is) throws IOException {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        byte[] internalBuffer = new byte[1024];
+        int read = is.read(internalBuffer);
+        while (read != -1) {
+            baos.write(internalBuffer, 0, read);
+            read = is.read(internalBuffer);
+        }
+        return baos.toByteArray();
+    }
+
+    protected URL createAPIKeyUrl() throws MalformedURLException, URISyntaxException {
+        return new URL(defaultUrl, (defaultUrl.toURI().getPath() + "/_security/api_key").replaceAll("/+", "/"));
+    }
+
+    protected URL getHttpInfoUrl() throws MalformedURLException, URISyntaxException {
+        return new URL(defaultUrl, (defaultUrl.toURI().getPath() + "/_nodes/http").replaceAll("/+", "/"));
+    }
+
+    protected static List<String> getBoundAddresses(Map<?, ?> nodesInfo) {
+        nodesInfo = (Map<?, ?>) nodesInfo.get("nodes");
+        Map<?, ?> nodeInfo = (Map<?, ?>) nodesInfo.values().iterator().next();
+        Map<?, ?> http = (Map<?, ?>) nodeInfo.get("http");
+        return (ArrayList<String>) http.get("bound_address");
+    }
+
+    static String getVersion(Map<?, ?> nodesInfo) {
+        nodesInfo = (Map<?, ?>) nodesInfo.get("nodes");
+        Map<?, ?> nodeInfo = (Map<?, ?>) nodesInfo.values().iterator().next();
```

Comment:
I **propose** we be strict here, and error out if enrolling in a mixed version cluster.
We haven't discussed supporting creating the enroll token on a node and using it on another node.
Your call.

Reply:
I think the idea was that some versions may be compatible while some not, but I agree it could be confusing unless we support  some kind of compatibility matrix.

Reply:
As we discussed today, I am changing to get only current node info